r/tests/testthat/test-dplyr-string-functions.R

```diff
@@ -719,6 +719,69 @@ test_that("errors in strptime", {
   )
 })
 
+test_that("strftime", {
+  skip_on_os("windows") # https://issues.apache.org/jira/browse/ARROW-13168
+
+  # TODO: consider reevaluating this workaround after ARROW-12980
+  withr::local_timezone("UTC")
+  times <- tibble(x = c(lubridate::ymd_hms("2018-10-07 19:04:05"), NA))
+  seconds <- tibble(x = c("05.000000", NA))
+  formats_minus_c <- "%a %A %w %d %b %B %m %y %Y %H %I %p %M %z %Z %j %U %W %x %X %% %G %V %u"
+  formats <- paste(formats_minus_c, "%c")
+
+  expect_dplyr_equal(
+    input %>%
+      mutate(x = strftime(x, format = formats)) %>%
+      collect(),
+    times
+  )
+
+  withr::with_timezone("Pacific/Marquesas",
+   expect_dplyr_equal(
+      input %>%
+        mutate(x = strftime(x, format = formats)) %>%
+        collect(),
+      times
+    )
+  )
+
+  expect_dplyr_equal(
+    input %>%
+      mutate(x = strftime(x, format = formats_minus_c, tz = "Pacific/Marquesas")) %>%
+      collect(),
+    times
+  )
+
+  withr::with_locale(new = c("LC_TIME" = "C"),
+    expect_dplyr_equal(
+      input %>%
+        mutate(x = strftime(x, format = "%c", tz = "Pacific/Marquesas")) %>%
+        collect(),
+      times
+    )
+  )
```

Comment:
It appears R's strftime interprets `%c` differently as `date.h` we use in arrow in certain locales and timezones?
Sometimes it's `"%a %b %e %H:%M:%S %Y"` and sometimes `"%a %b %e %H:%M:%S %Y %Z"`.
Could also just be my laptop.

Reply:
I'm a bit confused here - can you elaborate a bit more on your comments above please?

Reply:
Sorry, it was late :).
These tests are failing locally for me on the `%c` flag in `Pacific/Marquesas`, but not for say `UTC`. I am not sure why - it could be a difference between the definition of `%c` in [`date.h`](https://howardhinnant.github.io/date/date.html#to_stream_formatting) we're using and in R's strftime. Or it could just be some setting on my laptop. Either way it's very strange to see an error on only this flag with this locale and timezone.

I'm more and more suspicious of my laptop really.

Reply:
Another possibility is that the timezone database your system uses that may/may not be quite the same in different environments (cf https://stat.ethz.ch/R-manual/R-devel/library/base/html/timezones.html)

Reply:
Could be. FWIW I'm on Ubuntu 21.04 and have `LC_TIME=nl_NL.UTF-8` and `Sys.timezone()->  "Europe/Amsterdam"`.
I'll remove the check and see what happens in the CI.
Should we add checks in for different locales into the tests?

Reply:
If we're worried that there might be differences / bugs not respecting locales, having something non-default using  `withr::with_locale` (like you do above) would be a good idea. 

[We do hardcode this for at least some of our CI builds](https://github.com/apache/arrow/blob/master/ci/docker/linux-apt-r.dockerfile#L113)

Reply:
I've disabled `%c` in R for now and I propose we merge this and revisit later.
@pitrou shall we block %c in c++ as well? See https://github.com/HowardHinnant/date/issues/704